src/s3/async_put_object.rs

```diff
@@ -106,12 +106,8 @@ impl<'a> AsyncWrite for AsyncPutObject<'a> {
                     .body(mem::take(&mut self.buf).into())
                     .acl(ObjectCannedAcl::BucketOwnerFullControl)
                     .send();
-                self.state = PutObjectState::Closing(Box::pin(fut));
-                cx.waker().wake_by_ref();
```

Comment:
GitHub has made a meal of this diff, but essentially I'm replacing this call to `wake_by_ref` with an explicit poll of the future, then transitioning directly to `Closed` if it completes straight away or `Closing` otherwise.

I think this is the key - since we're returning `Pending` in this control flow branch, we have to arrange to be woken up *somehow*. Waking ourselves works, but polling the future also seems to do the trick AFAICT.